rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java

```diff
@@ -0,0 +1,334 @@
+                Set<J> validContexts = contexts.get(instanceOf);
+                for (Iterator<?> it = cursor.getPath(); it.hasNext(); ) {
+                    Object next = it.next();
+                    if (validContexts.contains(next)) {
+                        replacements.put(typeCast, instanceOf);
+                        break;
+                    } else if (root == next) {
+                        break;
+                    }
+                }
+            }
+        }
+
+        public boolean isEmpty() {
+            return replacements.isEmpty();
+        }
+
+        public J.InstanceOf processInstanceOf(J.InstanceOf instanceOf, Cursor cursor) {
+            @Nullable JavaType type = toJavaType((TypeTree) instanceOf.getClazz());
+            String name = patternVariableName(instanceOf, cursor);
+            J.InstanceOf result = instanceOf.withPattern(new J.Identifier(
+                    randomId(),
+                    Space.build(" ", emptyList()),
+                    Markers.EMPTY,
+                    name,
+                    type,
+                    null));
+
+            // update entry in replacements to share the pattern variable name
+            for (Map.Entry<J.TypeCast, J.InstanceOf> entry : replacements.entrySet()) {
+                if (entry.getValue() == instanceOf) {
+                    entry.setValue(result);
+                }
+            }
+            return result;
+        }
+
+        @Nullable
+        public J processTypeCast(J.TypeCast typeCast, Cursor cursor) {
+            J.InstanceOf instanceOf = replacements.get(typeCast);
+            if (instanceOf != null && instanceOf.getPattern() != null) {
+                String name = ((J.Identifier) instanceOf.getPattern()).getSimpleName();
+                TypedTree owner = cursor.firstEnclosing(J.MethodDeclaration.class);
+                owner = owner != null ? owner : cursor.firstEnclosingOrThrow(J.ClassDeclaration.class);
+                JavaType.Variable fieldType = new JavaType.Variable(null, Flag.Default.getBitMask(), name, owner.getType(), typeCast.getType(), emptyList());
+                return new J.Identifier(
+                        randomId(),
+                        typeCast.getPrefix(),
+                        Markers.EMPTY,
+                        name,
+                        typeCast.getType(),
+                        fieldType);
+            }
+            return null;
+        }
+    }
+
+    @Nullable
+    private static JavaType toJavaType(TypedTree typeTree) {
+        if (typeTree instanceof J.ArrayType) {
+            JavaType.Array result = new JavaType.Array(null, ((J.ArrayType) typeTree).getElementType().getType());
+            for (int i = 0; i < ((J.ArrayType) typeTree).getDimensions().size() - 1; i++) {
+                result = new JavaType.Array(null, result);
+            }
+            return result;
+        }
+        return typeTree.getType();
```

Comment:
@sambsnyd For `J.ArrayType` the `TypedTree#getType()` implementation always returns the element type (e.g. `int` for `int[][]`). Intuitively I would have expected the returned `JavaType` to also represent `int[][]`. But assuming it is implemented correctly, I implemented my own conversion here, so that multi-dimensional array types get compared correctly in the context of this recipe.

Reply:
`J.ArrayType` represents an appearance of a type in code, whereas `JavaType.Array` is metadata about that type. 
Consider this variable declaration which includes both:
`String[] s;`

There is a `J.ArrayType` for the "String[]" part of the declaration, which contains the whitespace and markers and such and is rendered in text. If you were to call `J.Identifier.getType()` on that identifier "s" it would return a `JavaType.Array` representing an array of Strings. `JavaType.Array` is metadata and does not literally appear in the text

I agree that this is confusing because these are obviously similar concepts. If you were to `null` out the type information, the `JavaType.Array`, you'd still get the same text printed out in the end. If you were to `null` out the `J.ArrayType` you'd print out the ` s;`, which is obviously not valid Java. Hopefully that helps to clarify the relationship.

Reply:
That part is quite clear to me. That is no different than in other ASTs I've been working with. Well, there is one difference: Normally the AST wouldn't already contain the types, since the AST was produced by a parser, rather than a compiler, as in this case.

Sorry about not explaining my question more clearly.